### PR TITLE
Pretty print `carton test` output

### PR DESCRIPTION
**Pull request by carson-katri**
Generalized the `ProcessRunner` to take a `ProcessOutputParser` (currently only `DiagnosticsParser` and `TestsParser` conform).

Here's a sample of passing & failing test cases:

Screen Shot 2020-10-26 at 5 51 19 PM
Screen Shot 2020-10-26 at 6 01 31 PM

And here's what the raw output of `swift test` would be:

```
Test Suite 'All tests' started at 2020-10-26 21:33:57.453
Test Suite 'testBundle.xctest' started at 2020-10-26 21:33:57.453
Test Suite 'Test' started at 2020-10-26 21:33:57.454
Test Case 'Test.testTrivial' started at 2020-10-26 21:33:57.460
/Users/carson/.carton/carton/TestApp/Tests/Tests/Test.swift:21: error: Test.testTrivial : XCTAssertEqual failed: ("Hello, world") is not equal to ("Hello world") - 
/Users/carson/.carton/carton/TestApp/Tests/Tests/Test.swift:22: error: Test.testTrivial : XCTAssertGreaterThan failed: ("5") is not greater than ("5") - 
/Users/carson/.carton/carton/TestApp/Tests/Tests/Test.swift:23: error: Test.testTrivial : XCTAssertLessThan failed: ("5") is not less than ("4") - 
/Users/carson/.carton/carton/TestApp/Tests/Tests/Test.swift:24: error: Test.testTrivial : XCTAssertLessThanOrEqual failed: ("5") is greater than ("4") - 
/Users/carson/.carton/carton/TestApp/Tests/Tests/Test.swift:25: error: Test.testTrivial : XCTAssertGreaterThanOrEqual failed: ("5") is less than ("6") - 
Test Case 'Test.testTrivial' failed (0.001 seconds)
Test Suite 'Test' failed at 2020-10-26 21:33:57.461
	 Executed 2 tests, with 6 failures (1 unexpected) in 0.007 (0.007) seconds
Test Suite 'testBundle.xctest' failed at 2020-10-26 21:33:57.461
	 Executed 2 tests, with 6 failures (1 unexpected) in 0.007 (0.007) seconds
Test Suite 'All tests' failed at 2020-10-26 21:33:57.461
	 Executed 2 tests, with 6 failures (1 unexpected) in 0.007 (0.007) seconds
```

### CODE REVIEW

**Review by MaxDesiatov — APPROVED** [review]
This is marvelous! 👏 I myself lost all hope for parsing XCTest output long time ago, and that's the reason I hacked an XCTest JSON output formatter in our development snapshots (but it didn't land in our 5.3 due to lack of time). This amazing work means I can revert the JSON output formatter chances, as it had almost no chances to land upstream in its current form, even though XCTest output in JSON format is more elegant in theory.

Maybe this output parser could become a separate library at some point? But I'm surely fine with merging it now here and battle-testing for a bit to make sure we can publish it later as a separate library for other people to rely on.

File: Sources/CartonHelpers/Parsers/ProcessOutputParser.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+public protocol ProcessOutputParser {
+  /// Parse the output of a `Process`, format it, then output in the `InteractiveWriter`.
+  func parse(_ output: String, _ terminal: InteractiveWriter)
+  var onlyOnFail: Bool { get }
```
Comment (MaxDesiatov) [inline Sources/CartonHelpers/Parsers/ProcessOutputParser.swift:18]: Please add a doc comment for this property.

**Comment by carson-katri** [comment]
Yeah it'd be nice to separate these from carton. Hopefully we can get Json output in the future (and for compiler output too 🤞)

**Comment by MaxDesiatov** [comment]
The compiler can already output parseable diagnostics, I'm not sure about JSON, but bitstream output is available for sure, and now can be parsed in pure Swift thanks to https://github.com/apple/swift-tools-support-core/pull/120 and https://github.com/apple/swift-tools-support-core/pull/145

**Comment by carson-katri** [comment]
Oh, awesome!

**Comment by carson-katri** [comment]
I tried to implement the DiagnosticsParser with `TSCUtility.SerializedDiagnostics`, but it seems to fall apart if there are fixits... I kept getting the error `unexpectedSubblock`.

**Review by MaxDesiatov — APPROVED** [review]
Thanks 👍

**Comment by MaxDesiatov** [comment]
Seems like TSC support for diagnostics in the bitcode format is unfinished 😕

**Review by yonihemi — APPROVED** [review]

**Pull request merged**